iron-input.html

```diff
@@ -127,11 +158,46 @@
     _bindValueChanged: function() {
       if (this.value !== this.bindValue) {
         this.value = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue;
+        this._updateBindValues();
       }
       // manually notify because we don't want to notify until after setting value
       this.fire('bind-value-changed', {value: this.bindValue});
     },
 
+    _bindValueAsNumberChanged: function() {
+      var newValue = this.bindValueAsNumber;
+      if ((this.type === 'number' || this.type === 'date') && this.valueAsNumber !== newValue) {
+        try {
```

Comment:
Please add a comment, though. I promise you 2 months from now we won't remember which way `<input type="number">` is quirky, and we'll stare at this for a long time.